Make Persistent Menu Navigation a Sidebar

AddisonDunn opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

Jira: [one](https://dimagi.atlassian.net/browse/USH-4587?atlOrigin=eyJpIjoiYTliNmZjY2Q2NWFkNGE0YjgyODM2MGQ4ZTBjNjA1OWEiLCJwIjoiaiJ9) and [two](https://dimagi.atlassian.net/browse/USH-4859).

This PR:
- Turns the persistent menu navigation into a sidebar when on desktop, but remaining an overlay on mobile
- To accomplish the bullet above, I needed to "lock" the main content container, so that scrolling happens on the content container instead of the entire page. This way the persistent menu and the other content can be separately scrollable.
- Removes the data preview bar on mobile
- Removes the gray background around forms on mobile

To see what I mean by "locked" and "separately scrollable", see this video:
https://github.com/user-attachments/assets/eed7edaf-1036-4fb0-a705-16e4f3847a57

Gray background removed on mobile:
![Screenshot 2024-09-18 at 4 03 01 PM](https://github.com/user-attachments/assets/4f07b0ae-cca2-414f-8cd3-901aeebeb60c)

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->

I made some key changes to margins and the underlying DOM layout of web apps that I am hoping will be more of a boon than a burden.

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

- Everything looks good with testing locally: main app selection screen, case list screen with case tiles and map, form screen
- This will go through QA

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

Going to go through QA, ticket TBD

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [X] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [X] Risk label is set correctly
- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change

AddisonDunn committed:
remove data preview bar on mobile

AddisonDunn committed:
move content padding to more intentional place

AddisonDunn committed:
rename main container

AddisonDunn committed:
make toolbar heights accessible variables

AddisonDunn committed:
make persistent menu nav a sidebar on desktop but still an overlay on mobile

AddisonDunn committed:
reorganize main layout so that main content container is 'locked'

and so that the sidebar persistent menu and main content container can be seperately scrollable, similar to how they are in Jira

AddisonDunn committed:
make breadcrumb bottom shadow a border instead

AddisonDunn committed:
remove form margins on mobile

AddisonDunn committed:
keep menu closed between pages if user closed it

nospame reviewed (commented):
Did you consider leaving the persistent nav as a regular Bootstrap offcanvas element and explicitly pushing over the main content on larger screens? Bootstrap provides options for the offcanvas both to not darken the background and to allow scrolling under it that could help make it possible. ([demo](https://getbootstrap.com/docs/5.0/components/offcanvas/#backdrop) and [options](https://getbootstrap.com/docs/5.0/components/offcanvas/#options))

The current implementation seems like it will probably work, but it's a pretty good chunk of code that might be mostly doable with built-in Bootstrap functionality.

esoergel commented on corehq/apps/cloudcare/templates/cloudcare/partials/menu/persistent_menu.html:
```diff
@@ -1,35 +1,32 @@
-    <div class="offcanvas-header persistent-menu-max-width d-flex align-items-center justify-content-between">
+  <div id="persistent-menu-container" class="bg-white h-100 position-relative overflow-scroll"
+    aria-labelledby="persistent-menu-label">
+    <button type="button" id="close-button" class="btn-close text-reset position-absolute"
```
`#close-button` might be too generic a term, risking conflict with other elements on the page

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -1546,9 +1546,39 @@ hqDefine("cloudcare/js/formplayer/menus/views", [
         events: {
             'click #app-main': 'onClickAppMain',
         },
+        handleSmallScreenChange: function (smallScreenEnabled) {
```
I'd love to be able to approach this with something like this:
```html
<div id="persistent-menu-content" style="display: none;">
  ...
</div>
<div id="persistent-menu-small">
  <button data-bs-toggle="offcanvas">...</button>
  <div class="offcanvas">
    {{ persistent-menu-content }}
  </div>
</div>
<div id="persistent-menu-big">
  <button data-bs-toggle="collapse">...</button>
  <div class="collapse show">
    {{ persistent-menu-content }}
  </div>
</div>
```
Where the containing context is swapped depending on which screen size is active, or perhaps the same DOM is displayed in two places, and we use media queries to figure out which to hide and which to show.  I don't know a good way to do that though.  Had the same problem with the login-as banner/pill.

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -1562,15 +1592,6 @@ hqDefine("cloudcare/js/formplayer/menus/views", [
             FormplayerFrontend.trigger("persistentMenuSelect");
         },
-        onBeforeDetach: function () {
-            // Be sure to hide offcanvas element so scroll works properly
```
is this not an issue anymore?

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -1546,9 +1546,39 @@ hqDefine("cloudcare/js/formplayer/menus/views", [
         onRender: function () {
             this.showChildView('menu', new PersistentMenuListView({collection: this.collection}));
         },
+        onDomRefresh: function () {
+            this.handleSmallScreenChange(cloudcareUtils.smallScreenIsEnabled());
```
out of curiosity, why is this in `onDomRefresh` rather than `onRender` or `initialize`?

esoergel commented on corehq/apps/cloudcare/templates/cloudcare/formplayer_home.html:
```diff
@@ -102,27 +102,31 @@
       <div class="d-lg-none">
         <div id="mobile-restore-as-region"></div>
       </div>
-      <div id="breadcrumb-region" class="print-container sticky-top"></div>
```
Does this make breadcrumbs no longer sticky on scroll?  Or is it that the whole top part of the screen (navbar and breadcrumbs) doesn't scroll at all?  With this change and with making everything a vertical flex element, does that mean we don't need to do the `.case-tile-container` and `.query-button-container` offset stuff anymore?

esoergel replied:
Ah, I see you did remove the offset for the latter already.

esoergel replied:
superb

esoergel commented on corehq/apps/cloudcare/templates/cloudcare/formplayer_home.html:
```diff
@@ -102,27 +102,31 @@
       </div>
-      <div class="d-lg-flex">
-        <div id="sidebar-region" class="noprint-sub-container mt-3"></div>
-        <div id="content-container" class="container mt-3">
```
this `container` is necessary to prevent wide elements from overflowing the div.  That was the issue addressed by https://github.com/dimagi/commcare-hq/pull/35069/commits/177ccc7b119a61df73d46fce3fb314cbd40999a7

I see you put a max-width on the element, which should probably address that, though does that also work well on smaller screen sizes?

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/app.js:
```diff
@@ -306,6 +307,7 @@ hqDefine("cloudcare/js/formplayer/app", [
         var sess = WebFormSession.WebFormSession(data);
         sess.renderFormXml(data, $('#webforms'));
         $('.menu-scrollable-container').addClass("d-none");
+        $('#sidebar-and-content').addClass('remove-padding-on-mobile');
```
Does this really have to apply only to form entry?  Wondering if we could instead remove the padding on mobile for everything.  Or if say menus do need padding, then maybe that could be part of the menu definition.

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -1561,6 +1561,9 @@ hqDefine("cloudcare/js/formplayer/menus/views", [
             } else {
                 $('#persistent-menu-container').removeClass(containerMobileClasses);
                 $('#persistent-menu-container').addClass(containerDesktopClasses);
+                if (sessionStorage.showPersistentMenu !== 'false') {
```
:star:

esoergel reviewed (commented):
Left a bunch of comments, mostly raising for discussion since I'm not certain how best to do a lot of this stuff - please don't take those as meaning "you must fix this."

nospame commented:
> 
>  This PR sort of doubles-down on that idea, by making the persistent menu _part_ of that flexbox layout instead of being an overlay. Since the desire of the request is to make the sidebar by default _shown_ and to "take up space" in the DOM, this seemed best to me, instead of giving the sidebar an `absolute` position and "pushing the content over" like we were previously doing for the SSCS sidebar.

This makes sense within the existing context, but I think it's worth considering/remembering that Bootstrap is really explicitly mobile-first. So by designing this as a web-first feature, this approach is basically working against Bootstrap instead of using it to our advantage. I don't know enough about the particular feature to say if I think it's the _wrong_ way to go about things here, but I think it's important to have the discussion about whether it is the _right_ approach (and maybe USH devs have already done that and I'm late to the party).

AddisonDunn committed:
remove unnecessary top margin for persistent case tile

it is offset by height of navbar and breadcrumbs automatically anyways

AddisonDunn committed:
use onAttach instead of onDomRefresh

earlier in the lifecycle, better practice

Jtang-1 commented on corehq/apps/cloudcare/templates/cloudcare/partials/form_entry/form.html:
```diff
@@ -47,7 +47,7 @@ <h1 class="title" data-bind="text: title, visible: !showInFormNavigation()"></h1
               css: { 'sticky-submit': isAnchoredSubmitStyle },
               style: {
-                'bottom':isAnchoredSubmitStyle && {{ request.couch_user.can_edit_data|yesno:'true,false' }} {% if environment != "web-apps" %} && false {% endif %} ? '30px' : '0' {# data preview bar #}
+                'bottom':isAnchoredSubmitStyle && !isAppPreview && !smallScreenIsEnabled && {{ request.couch_user.can_edit_data|yesno:'true,false' }}? '30px' : '0' {# data preview bar #}
```
I believe I did that and I also didn't prefer it. BTW this is only applicable for the sticky submit button (Where the submit button is shown as a bar anchored to the bottom of the screen) But from what I recall, it was tricky and involved some overhauling since the submit button and data preview bar and not within the same container. Also, both the anchored submit button and data preview bar are positioned using `position:fixed` and I couldn't figure out a way to make fixed items "float" even if they're on the same z-index.

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/app.js:
```diff
@@ -396,6 +396,11 @@ hqDefine("cloudcare/js/formplayer/app", [
+        self.smallScreenListener = CloudcareUtils.smallScreenListener(smallScreenEnabled => {
+            $.publish('smallScreenIsEnabled', smallScreenEnabled);
+        });
+        self.smallScreenListener.listen();
```
Do we need to also handle `stopListening` somewhere?

AddisonDunn committed:
change way that we modify sticky button bottom css based on debugger presence

AddisonDunn committed:
rename breadcrumb height var

AddisonDunn committed:
remove unused vars

AddisonDunn committed:
turn arrow toggle into a button

AddisonDunn committed:
give button a more complex ID

lower risk of name conflicts

AddisonDunn committed:
get persistent case tile working with new layout

bottom margin should only be added if persisent case tile is active. Otherwise the bottom margin creates blank space at the top of the form on mobile (and I don't think it should be there when not being used anyway). Could also be fixed with a flexbox but this solution seems simple enough and I think we have enough flexboxes at this point.

AddisonDunn committed:
hide persistent menu when printing

avoids weird blank space at top of printed form

nospame commented on corehq/apps/cloudcare/templates/cloudcare/partials/menu/persistent_menu.html:
```diff
@@ -1,35 +1,32 @@
 
 <script type="text/template" id="persistent-menu-template">
-  <a type="button"
+  <a id="persistent-menu-arrow-toggle"
+     type="button"
```
It's a little bit of a style preference, but semantic HTML is always nicer in my book.

FWIW `type="button"` is probably not doing anything here (https://stackoverflow.com/questions/25838318/is-it-alright-to-have-an-anchor-type-button) but if making it a `<button>` causes any issues I think the accessible way would be to give it the attribute `role="button"`.

AddisonDunn committed:
rename class

AddisonDunn committed:
put next-error button above sticky submit button

AddisonDunn committed:
fix next-error scrolling to adjust for new scrollable container

AddisonDunn committed:
adjust overflow when printing so entire form gets printed

AddisonDunn committed:
add role=button to see if makes accessible to NVDA screen reader

nospame commented on corehq/apps/hqwebapp/static/cloudcare/scss/formplayer-webapp/menu.scss:
```diff
@@ -70,3 +70,9 @@ $persistent-menu-image-size: 3rem;
   }
 }
+
+@media print {
+  #persistent-menu-region {
+    display: none;
```
You can also use `d-print-none` as a class on the element itself for this.

nospame reviewed (commented):
Just leaving more nitpicks on html/css, overall I do believe this looks good but haven't re-reviewed thoroughly 😅

AddisonDunn committed:
use d-print-none class

AddisonDunn committed:
use z-index variable

nospame commented on corehq/apps/cloudcare/templates/cloudcare/partials/menu/persistent_menu.html:
```diff
@@ -2,7 +2,8 @@
 <script type="text/template" id="persistent-menu-template">
   <button id="persistent-menu-arrow-toggle"
-     href="#"
+     role="button"
```
That's bizarre behavior from NVDA if so, but I believe it.

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/formplayer/menus/views.js:
```diff
@@ -616,6 +616,12 @@ hqDefine("cloudcare/js/formplayer/menus/views", [
             }
         },
+        onAttach: function () {
+            $('.case-tile-container').css('margin-bottom', '12px');
```
(not blocking) this seems hard to track down, since in the DOM inspector it'll just show up as an attribute of the element - does the margin have to go on the `.case-tile-container` or would it also work on the `#persistent-case-tile`?  If so, this could be defined by another class in `className` above, I think.

esoergel commented on corehq/apps/hqwebapp/static/cloudcare/scss/formplayer-webapp/content.scss:
```diff
@@ -35,8 +34,23 @@ body {
 }
 
+#content-plus-persistent-menu-container {
+  height: calc(100vh - #{$cloudcare-nav-height});
```
that's fancy

AddisonDunn committed:
use different way of specifying margin for pers. case tile
